Adds GPU CUDA kernel for scatter ops

fpmchu opened the pull request:
This adds GPU CUDA kernels for scatter_add, scatter_sub, and scatter_update.

Testing on a local machine with a weak graphics card already shows good improvement.  For about 20M floats, split as a 2Mx10 matrix, 1M size for idx, the op on a weak 3-core GPU is about twice as fast as on a 12-core CPU.  On very small data sets (1000 floats) the GPU code seems a bit slower (20%).

Some notes:
1. range checking in CUDA kernel currently not implemented.  It's not possible to include `OpKernelContext` since it includes protos hence nvcc fails to compile.  I can do it as a method in the .h/.cc files.  However, it seems rather inefficient as for low-dimensional input it is basically as slow as doing the entire scatter op sequentially on CPU.  Would it make sense to run the entire kernel, ignore bad indices, and report after?  This does mean the parameters may be updated when the code does fail because the modifications are happening in place.  Advice?<br>
2. There's a weird bug in `scatter_op_gpu.cu.cc` where I cannot use `TF_CALL_GPU_NUMBER_TYPES`.  I suspect it has to do with the `defined(__ANDROID__)` magic in `register_types.h` interacting badly with nvcc, but cannot confirm.

tensorflow-jenkins commented:
Can one of the admins verify this patch?

vrv commented:
There are a lot of unrelated commits in this PR -- can you separate them out into individual PRs?  Specifically, fixing the typos and such can be done as a separate PR, and then the main GPU impl + test can be in a single commit for reviewing.

fpmchu commented:
Done.

fpmchu commented:
Btw, with this change, I could re-enable word2vec example GPU support: https://github.com/fpmchu/tensorflow/commit/121784b62da64f3a799cc0eb6776ca4f35594a9f

However, on my machine, this made the example *slower*.  Either it's because my card is weak, or perhaps the example needs to be optimized for using GPU?  I'm not sure if it's my code that's slow...

martinwicke commented:
Quite possibly adding this adds a bunch of transfers back and forth, which
are slow. Optimizing the example for GPU would fix that.

Martin
On Fri, Jan 15, 2016 at 18:52 Frank Chu <notifications@github.com> wrote:

> Btw, with this change, I could re-enable word2vec example GPU support:
> fpmchu@121784b
> <https://github.com/fpmchu/tensorflow/commit/121784b62da64f3a799cc0eb6776ca4f35594a9f>
>
> However, on my machine, this made the example *slower*. Either it's
> because my card is weak, or perhaps the example needs to be optimized for
> using GPU? I'm not sure if it's my code that's slow...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/782#issuecomment-172149174>
> .
>

keveman commented on tensorflow/core/kernels/scatter_op.h:
```diff
@@ -0,0 +1,75 @@
+  template <typename Params, typename Update>
+  EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE
+  static void Run(Params&& p, const Update& u) {
+    p += u;
```
I am not sure about your intention of making the type of 'p' an rvalue reference. When called from the CPU kernel, Params would be a Eigen expression passed by value, so the `+=` would mutate the lhs. But from your GPU kernel below, you are passing in `params[params_i]` by value. How do you expect that to be updated? Also, the following is the semantics of scatter_add : 
> Duplicate entries are handled correctly: if multiple indices reference the same location, their contributions add.

This code when called from many threads concurrently will lose most of the updates. Please consider reading up on how to scatter operations on cuda. [Thrust](https://github.com/thrust/thrust/blob/master/testing/backend/cuda/scatter.cu) has some examples.

keveman replied:
I do understand that `p` in `Params&& p` is really a [universal reference](https://isocpp.org/blog/2012/11/universal-references-in-c11-scott-meyers) and not an rvalue reference. However, it's just not common usage in rest of TF, so I would request you just use a pointer.

fpmchu replied:
I used universal reference because CPU code does
```params.template chip<0>(indices(i)) = updates.template chip<0>(i)```
where the LHS is a rvalue.  If I used a pointer, I would have to first store the LHS in a temporary variable, then pass its address.  Besides the code looking ugly, it may also be less efficient.

It felt like universal reference is designed to solve exactly this problem.  But, there's a simple workaround, which is to store the LHS in a temporary variable, then use plain old references.  I did this now.  PTAL.

I will look into the thread-safety next.  Thanks for pointing to Thrust.  I also realized how to do range checks efficiently now without temporary overwriting data - just write a fast range check kernel.  I'll do that as well.

fpmchu commented:
@keveman After reading more, it seems like Thrust's scatter does not support repeated items.  [Thrust's doc](https://thrust.github.io/doc/group__scattering.html) says:
> If the same index appears more than once in the range [map, map + (last - first)), the result is undefined
Looking at [their CUDA code](https://github.com/thrust/thrust/blob/67bbec34a7fdb9186aea6c269870b517e42e4388/thrust/system/detail/generic/scatter.inl), I don't think they do anything special to handle repeated indices either.

Similarly, numpy also says the behavior is undefined.  From [numpys ndarray's doc](http://docs.scipy.org/doc/numpy/reference/arrays.indexing.html)
> For advanced assignments, there is in general no guarantee for the iteration order. This means that if an element is set more than once, it is not possible to predict the final result
In fact, as written, ```scatter_ops_test.py``` would fail right now when repeated indices exist.  For example,
```
$ n = np.array([[1, 2], [3, 4], [5, 6]])
$ n[np.array([1, 1])] += np.array([[10, 10], [20, 20]])
$ print(n)
[[ 1  2]
 [23 24]
 [ 5  6]]
```

Now of course, I agree this should be supported, if only because it's already written into the contract.  I just want to point it out since I would have to make some changes to the test as well.

keveman commented:
@fpmchu, I apologize, I should have checked Thrust's documentation before sending you on a wild goose chase. But you are right, since we have committed to this in the documentation, it'll be nice to support repeated indices. My belief is, the performance may not be terrible if you turned the `+=` to `atomicAdd`. Do you mind giving it a shot and running the benchmarks?

fpmchu commented:
Ok, added atomic ops and range check.

For atomics, the benchmark code as is doesn't really run on GPU (not that I can tell).  I wrote a custom script to benchmark, and it shows very little difference in speed between the non-atomic and the atomic version, for the common case when there are no repeats.  I think this is to be expected as the atomics are lock free.  It might be slow for when there are repeat indices, but that really seems like a odd use case anyway...

I had to add atomicAdd/atomicExch to support double, because CUDA does not support them directly.  The code is copied straight from CUDA's own manual.  I put them in ```cuda_kernel_helper.h```.  Not sure if that's a good idea and/or it's the right place.

For range checks, I wrote a simple CUDA kernel for doing range checks, and also added tests.

Btw, I'm not checking for CUDA errors, and it seems like I should.  I'm following examples from bias_ops though, and in there there's also no error checking... Any suggestions?

keveman commented:
fpmchu@, thanks for doing this, this is awesome! Almost there, we'll merge this as soon.

fpmchu commented:
I fixed the ```atomicAdd``` ovelroad.  PTAL.
However, I did consider it before, and I felt overloading might be slightly better because that forces a compilation error when one compiles against a new CUDA version.  This means the code will be forced to be fixed, instead of using an inefficient version.  That said, a custom wrapper isn't that bad so I did it anyway.

fpmchu commented:
As for ```scatter_update```, sorry that I didn't check the doc on it.  No, the current version will not implement
> If indices contains duplicate entries, lexicographically later entries override earlier entries.

It'll be quite tricky to do that.  Off the top of my head, I guess one can first sort then remove duplicates (or do a max).  It'll be inefficient and is tailored to a case that probably won't be used.  How about changing documentation?

fpmchu commented:
Two more questions
1. In my implementation I'm using template specialization.  I could easily use function overload.  They behave the same way I think, so not sure what's preferred.
2. Any comment on error checking?

keveman commented:
* Changing the documentation for `scatter_update` to note that one of the values corresponding to the duplicate indices will be written at that index.
* The compiler looks for function overloads before it looks for full specializations, so I usually write overloads. But in this case, both function overload and full specialization should behave the same, so I am leaving it up to you.

fpmchu commented:
Done and done.
After talking to @martinwicke , he says just to change doc in state_ops.cc.  Tests pass locally.  Anything else?

martinwicke commented:
Jenkins, test this please.

martinwicke commented on tensorflow/core/ops/state_ops.cc:
```diff
@@ -182,8 +182,8 @@ This operation computes
-If `indices` contains duplicate entries, lexicographically later entries
-override earlier entries.
+If `indices` contain multiple entries that reference the same location, any
+entry could be used to do the update; the choice is not defined.
```
maybe: 

If `indices` contains multiple entries that reference the same location in `updates`, the output could be updated to any one of the referenced values. Which value is used is undefined.

fpmchu replied:
Done.

fpmchu commented:
Added a note in RELEASE.md.  PTAL.

girving commented:
Can you smash all of these commits into one large commit?

fpmchu commented:
Done.

girving commented:
Thanks, looking through this now.

girving commented:
Can you do some merging of the registration macros in `scatter_op.cc`?  Certainly add and sub always have the same set of types, and for gpu it looks like everything has the same set of types.  Taking advantage of this would dramatically shorten that section and make it easier to read and modify.  For example, there would be only one `#if GOOGLE_CUDA` line in `scatter_op.cc`.

girving commented:
I'm likely being a bit dense here, but what file defines the GPU implementations?  I can't find anything in your new single CL.

Once I do find it: we may need to be more specific about what happens for `ScatterUpdate` on larger types.  Currently you guarantee that one of the indices is used (which admittedly is what I said to do), but that's not remotely actually guaranteed for either indices which refer to slices or even to `complex128` on CPU if we were to switch to threads.

girving commented:
Also, not sure if `complex64` is an issue yet, but once it is we'll want a primitive which is slightly weaker than atomic add / sub, since it's fine to add both components at unrelated points in the sequence.

fpmchu commented:
Sorry I missed some files during the squash.  Re-pushed now including the kernel.

fpmchu commented:
Hi, I made several changes based on comments.  PTAL.

Commit:
Adds GPU CUDA kernel for scatter ops.

fpmchu commented:
Done.  I also rebased to master to resolve some conflicts.

zheng-xq commented:
@vrv, @martinwicke, @keveman, @girving, the GPU portion of the code LGTM. 

Please feel free to move forward if you think other parts are good as well.

martinwicke commented:
Jenkins, test this please.

martinwicke commented:
Jenkins, test this please.

tensorflow-jenkins commented:
Can one of the admins verify this patch?

fpmchu commented:
Hi, sorry for delay.
I've fixed the test by changing ```testScatterOutOfRange```  (please check out lines 140-165).
It's a bit tricky because I have to detect whether GPU exists or not.

Can you please test again?

fpmchu commented:
I have fixed the test, and rebased + squashed.  Please test again

martinwicke commented:
Jenkins, please test this.

martinwicke commented:
Jenkins, test this please

martinwicke commented:
Hold on, this may be a tool failure.

martinwicke commented:
Jenkins, test this please.

fpmchu commented:
After talking with @martinwicke, we've decided to temporarily disabled the GPU range check test because gpu_pip test does not run on a GPU and is failing.

Squashed and rebased.  PTAL.

martinwicke commented:
Jenkins, test this please.
